File: BOJ/java/16194.java
import java.util.Arrays;
import java.util.Scanner;

class Main {
    static int[] d = new int[1001];
    static int[] p = new int[1001];
    static int maxValue = 1000 * 10000;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();

        for (int i = 1; i <= n; i++) {
            p[i] = sc.nextInt();
        }
        Arrays.fill(d, maxValue);

        sc.close();
        System.out.println(go(n));
    }

    static int go(int i) {
        if (i == 0) return 0;
        if (d[i] != maxValue) return d[i];

        for (int j = 1; j <= i; j++) {
            d[i] = Math.min(d[i], go(i - j) + p[j]);
        }

        return d[i];
    }
}

File: BOJ/kotlin/16194.kt
import java.util.*

val maxValue = 1000 * 10000
val d = IntArray(1001) { maxValue }
val p = IntArray(1001) { 0 }

fun main() {
    val sc = Scanner(System.`in`)
    val n = sc.nextInt()
    for (i in 1..n) p[i] = sc.nextInt()
    println(go(n))
}

fun go(i: Int): Int {
    if (i == 0) return 0
    if (d[i] != maxValue) return d[i]
    for (j in 1..i) d[i] = kotlin.math.min(d[i], go(i - j) + p[j])
    return d[i];
}

